File: frontend/src/stores/quiz.ts
import { defineStore } from 'pinia';
import { ref, computed } from 'vue';
import type { QuizQuestion, QuizAttempt } from '@/types';
import { useApi } from '@/composables/useApi';

const api = useApi();

export const useQuizStore = defineStore('quiz', () => {
  // State
  const currentAttempt = ref<QuizAttempt | null>(null);
  const currentQuestion = ref<QuizQuestion | null>(null);
  const questions = ref<QuizQuestion[]>([]);
  const questionIndex = ref(0);
  const score = ref(0);
  const difficulty = ref(1);
  const isLoading = ref(false);
  const selectedOption = ref<number | null>(null);
  const showResult = ref(false);
  const isCorrect = ref(false);

  // Getters
  const progress = computed(() => {
    if (!currentAttempt.value) return 0;
    return (currentAttempt.value.total_questions / 10) * 100;
  });

  const streak = computed(() => currentAttempt.value?.streak_count || 0);

  // Actions
  async function startQuiz(topicId: number) {
    isLoading.value = true;
    try {
      // Create attempt
      const attemptRes = await api.createAttempt(topicId);
      currentAttempt.value = attemptRes.data;

      // Get questions
      const questionsRes = await api.getQuestions(topicId, difficulty.value);
      questions.value = questionsRes.data;
      currentQuestion.value = questions.value[0];
      questionIndex.value = 0;
      score.value = 0;
    } finally {
      isLoading.value = false;
    }
  }

  async function submitAnswer(optionIndex: number) {
    if (!currentAttempt.value || !currentQuestion.value) return;

    selectedOption.value = optionIndex;
    isLoading.value = true;

    try {
      const response = await api.submitAnswer({
        attempt_id: currentAttempt.value.id,
        question_id: currentQuestion.value.id,
        selected_option_index: optionIndex,
        time_taken_seconds: 30, // TODO: track actual time
      });

      isCorrect.value = response.data.is_correct;
      showResult.value = true;

      // Update attempt
      if (currentAttempt.value) {
        currentAttempt.value.total_questions = response.data.total_questions || currentAttempt.value.total_questions + 1;
        if (isCorrect.value) {
          currentAttempt.value.correct_answers++;
          currentAttempt.value.streak_count++;
        } else {
          currentAttempt.value.streak_count = 0;
        }
      }

      // Move to next question or finish
      setTimeout(() => {
        if (response.data.next_question) {
          currentQuestion.value = response.data.next_question;
          selectedOption.value = null;
          showResult.value = false;
          questionIndex.value++;
        } else {
          // Quiz complete
          currentQuestion.value = null;
        }
      }, 1500);
    } finally {
      isLoading.value = false;
    }
  }

  function resetQuiz() {
    currentAttempt.value = null;
    currentQuestion.value = null;
    questions.value = [];
    questionIndex.value = 0;
    score.value = 0;
    selectedOption.value = null;
    showResult.value = false;
    isCorrect.value = false;
  }

  return {
    currentAttempt,
    currentQuestion,
    questions,
    questionIndex,
    score,
    difficulty,
    isLoading,
    selectedOption,
    showResult,
    isCorrect,
    progress,
    streak,
    startQuiz,
    submitAnswer,
    resetQuiz,
  };
});

File: backend/app/api/v1/quiz.py
"""Quiz API endpoints."""
from typing import List
from uuid import UUID
from datetime import datetime

from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select, func

from app.database import get_async_session
from app.models.quiz import QuizQuestion, QuizAttempt, QuizResponse
from app.models.users import User
from app.schemas.quiz import (
    QuizQuestionResponse, QuizAttemptCreate, QuizAttemptResponse,
    QuizAnswerSubmit, QuizAnswerResponse, AdaptiveQuizRequest
)
from app.core.security import get_current_user
from app.services.ai_service import AIService

router = APIRouter(prefix="/quiz", tags=["quiz"])


@router.get("/topics/{topic_id}/questions", response_model=List[QuizQuestionResponse])
async def get_questions(
    topic_id: int,
    difficulty: int = 1,
    limit: int = 10,
    db: AsyncSession = Depends(get_async_session)
):
    """Get quiz questions for a topic."""
    result = await db.execute(
        select(QuizQuestion)
        .where(QuizQuestion.topic_id == topic_id)
        .where(QuizQuestion.difficulty_level == difficulty)
        .order_by(func.random())
        .limit(limit)
    )
    questions = result.scalars().all()
    return questions


@router.post("/attempts", response_model=QuizAttemptResponse)
async def create_attempt(
    attempt_data: QuizAttemptCreate,
    db: AsyncSession = Depends(get_async_session),
    current_user: User = Depends(get_current_user)
):
    """Start a new quiz attempt."""
    attempt = QuizAttempt(
        user_id=current_user.id,
        topic_id=attempt_data.topic_id
    )
    db.add(attempt)
    await db.commit()
    await db.refresh(attempt)
    return attempt


@router.post("/answer", response_model=QuizAnswerResponse)
async def submit_answer(
    answer: QuizAnswerSubmit,
    db: AsyncSession = Depends(get_async_session),
    current_user: User = Depends(get_current_user)
):
    """Submit an answer and get next question."""
    # Get question
    result = await db.execute(
        select(QuizQuestion).where(QuizQuestion.id == answer.question_id)
    )
    question = result.scalar_one_or_none()

    if not question:
        raise HTTPException(status_code=404, detail="Question not found")

    # Check if answer is correct
    is_correct = answer.selected_option_index == question.correct_option_index

    # Create response record
    response = QuizResponse(
        attempt_id=answer.attempt_id,
        question_id=answer.question_id,
        selected_option_index=answer.selected_option_index,
        is_correct=is_correct,
        time_taken_seconds=answer.time_taken_seconds
    )
    db.add(response)

    # Update attempt stats
    attempt_result = await db.execute(
        select(QuizAttempt).where(QuizAttempt.id == answer.attempt_id)
    )
    attempt = attempt_result.scalar_one()

    attempt.total_questions += 1
    if is_correct:
        attempt.correct_answers += 1
        attempt.streak_count += 1
    else:
        attempt.streak_count = 0

    # Calculate score percentage
    attempt.score_percentage = int((attempt.correct_answers / attempt.total_questions) * 100)

    # Adjust difficulty
    if attempt.streak_count >= 3 and attempt.difficulty_reached < 3:
        attempt.difficulty_reached += 1
    elif attempt.streak_count == 0 and attempt.difficulty_reached > 1:
        attempt.difficulty_reached -= 1

    await db.commit()

    # Get next question
    next_question_result = await db.execute(
        select(QuizQuestion)
        .where(QuizQuestion.topic_id == attempt.topic_id)
        .where(QuizQuestion.difficulty_level == attempt.difficulty_reached)
        .where(QuizQuestion.id != answer.question_id)
        .order_by(func.random())
        .limit(1)
    )
    next_question = next_question_result.scalar_one_or_none()

    return QuizAnswerResponse(
        is_correct=is_correct,
        correct_option_index=question.correct_option_index,
        explanation=question.explanation,
        next_question=next_question,
        attempt_complete=attempt.total_questions >= 10,  # Limit to 10 questions
        score_percentage=attempt.score_percentage
    )


@router.get("/attempts/{attempt_id}", response_model=QuizAttemptResponse)
async def get_attempt(
    attempt_id: UUID,
    db: AsyncSession = Depends(get_async_session),
    current_user: User = Depends(get_current_user)
):
    """Get quiz attempt details."""
    result = await db.execute(
        select(QuizAttempt)
        .where(QuizAttempt.id == attempt_id)
        .where(QuizAttempt.user_id == current_user.id)
    )
    attempt = result.scalar_one_or_none()

    if not attempt:
        raise HTTPException(status_code=404, detail="Attempt not found")

    return attempt


@router.post("/adaptive/generate")
async def generate_adaptive_question(
    request: AdaptiveQuizRequest,
    ai_service: AIService = Depends(),
    db: AsyncSession = Depends(get_async_session)
):
    """Generate adaptive quiz question using AI."""
    # Get topic info
    from app.models.subjects import Topic
    topic_result = await db.execute(
        select(Topic).where(Topic.id == request.topic_id)
    )
    topic = topic_result.scalar_one_or_none()

    if not topic:
        raise HTTPException(status_code=404, detail="Topic not found")

    # Generate question using AI
    question_data = await ai_service.generate_quiz_question(
        topic=topic.name,
        standard_code=None,
        difficulty=request.difficulty,
        previous_questions=[str(qid) for qid in request.exclude_question_ids]
    )

    return question_data
